src/jit/importer.cpp

```diff
@@ -7102,33 +7120,32 @@ GenTreePtr          Compiler::impFixupStructReturnType(GenTreePtr op, CORINFO_CL
 #if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
 
 #ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
-    // No VarArgs for CoreCLR.
-    assert(!info.compIsVarArgs); 
+    // No VarArgs for CoreCLR on x64 Unix
+    assert(!info.compIsVarArgs);
 
-    if (varTypeIsStruct(info.compRetNativeType))
+    // Is method returning a multi-reg struct?
+    if (varTypeIsStruct(info.compRetNativeType) && IsMultiRegReturnedType(retClsHnd))
     {
-        SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR structDesc;
-        eeGetSystemVAmd64PassStructInRegisterDescriptor(retClsHnd, &structDesc);
+        // In case of multi-reg struct return, we force IR to be one of the following:
+        // GT_RETURN(lclvar) or GT_RETURN(call).  If op is anything other than a
+        // lclvar or call, it is assigned to a temp to create: temp = op and GT_RETURN(tmp).
 
-        if (structDesc.passedInRegisters)
+        if (op->gtOper == GT_LCL_VAR)
         {
-            if (op->gtOper == GT_LCL_VAR)
-            {
-                // This LCL_VAR is a register return value, it stays as a TYP_STRUCT
-                // Make sure this struct type stays as struct so that we can return it in registers.
-                unsigned lclNum = op->gtLclVarCommon.gtLclNum;                
-                lvaTable[lclNum].lvIsMultiRegArgOrRet = true;
-
-                return op;
-            }
+            // Make sure that this struct stays in memory and doesn't get promoted.
+            unsigned lclNum = op->gtLclVarCommon.gtLclNum;
+            lvaTable[lclNum].lvIsMultiRegArgOrRet = true;
+            lvaTable[lclNum].lvDoNotEnregister = true;
```

Comment:
You have changed the semantics by adding the lvDoNotEnregister. Is that newly required for some reason? I think it would be better to do this as part of the struct promotion, which I think is where it is being handled now. In any case, you shouldn't be directly setting lvDoNotEnregister. You should call lvaSetVarDoNotEnregister with a "reason".